dbms/src/Operators/TopNTransformOp.cpp

```diff
@@ -0,0 +1,58 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <DataStreams/MergeSortingBlocksBlockInputStream.h>
+#include <Interpreters/sortBlock.h>
+#include <Operators/TopNTransformOp.h>
+
+namespace DB
+{
+OperatorStatus TopNTransformOp::transformImpl(Block & block)
+{
+    if (unlikely(!block))
+    {
+        if (!blocks.empty())
+        {
+            impl = std::make_unique<MergeSortingBlocksBlockInputStream>(
+                blocks,
+                order_desc,
+                req_id,
+                max_block_size,
+                limit);
+            block = impl->read();
+        }
+        return OperatorStatus::HAS_OUTPUT;
+    }
+    RUNTIME_CHECK_MSG(!impl, "Impl must be nullptr here.");
+
+    sortBlock(block, order_desc, limit);
+    blocks.emplace_back(std::move(block));
+    return OperatorStatus::NEED_INPUT;
+}
+
+OperatorStatus TopNTransformOp::tryOutputImpl(Block & block)
+{
+    if (impl)
```

Comment:
Here we can not distingish the following two cases:
1. the input is all read, but since `blocks` is empty, so `impl` is nullptr
2. there still need to read from the input, so `impl` is nullptr

is this ok?

Reply:
1. case 1 indicates that there are no blocks, then we don't need impl, just return HAS_OUTPUT is okay. I have added an Empty block test to ensure the safety.
2. case 2 is impossible. If the input is blocked, tryOutputImpl will not be called.